impeller/entity/entity_pass.cc

```diff
@@ -826,81 +818,12 @@ bool EntityPass::RenderElement(Entity& element_entity,
-      auto op = clip_coverage_stack.back().coverage;
-      clip_coverage_stack.push_back(
-          ClipCoverageLayer{.coverage = clip_coverage.coverage,
-                            .clip_depth = element_entity.GetClipDepth() + 1});
-      FML_DCHECK(clip_coverage_stack.back().clip_depth ==
-                 clip_coverage_stack.front().clip_depth +
-                     clip_coverage_stack.size() - 1);
-
-      if (!op.has_value()) {
-        // Running this append op won't impact the clip buffer because the
-        // whole screen is already being clipped, so skip it.
-        return true;
-      }
-    } break;
-    case Contents::ClipCoverage::Type::kRestore: {
-      if (clip_coverage_stack.back().clip_depth <=
-          element_entity.GetClipDepth()) {
-        // Drop clip restores that will do nothing.
-        return true;
-      }
-
-      auto restoration_index = element_entity.GetClipDepth() -
-                               clip_coverage_stack.front().clip_depth;
-      FML_DCHECK(restoration_index < clip_coverage_stack.size());
-
-      // We only need to restore the area that covers the coverage of the
-      // clip rect at target depth + 1.
-      std::optional<Rect> restore_coverage =
-          (restoration_index + 1 < clip_coverage_stack.size())
-              ? clip_coverage_stack[restoration_index + 1].coverage
-              : std::nullopt;
-      if (restore_coverage.has_value()) {
-        // Make the coverage rectangle relative to the current pass.
-        restore_coverage = restore_coverage->Shift(-global_pass_position);
-      }
-      clip_coverage_stack.resize(restoration_index + 1);
-
-      if constexpr (ContentContext::kEnableStencilThenCover) {
-        // Skip all clip restores when stencil-then-cover is enabled.
-        if (clip_coverage_stack.back().coverage.has_value()) {
-          clip_replay_->RecordEntity(element_entity, clip_coverage.type);
-        }
-        return true;
-      }
-
-      if (!clip_coverage_stack.back().coverage.has_value()) {
-        // Running this restore op won't make anything renderable, so skip it.
-        return true;
-      }
-
-      auto restore_contents =
-          static_cast<ClipRestoreContents*>(element_entity.GetContents().get());
-      restore_contents->SetRestoreCoverage(restore_coverage);
-
-    } break;
-  }
-
-#ifdef IMPELLER_ENABLE_CAPTURE
-  {
-    auto element_entity_coverage = element_entity.GetCoverage();
-    if (element_entity_coverage.has_value()) {
-      element_entity_coverage =
-          element_entity_coverage->Shift(global_pass_position);
-      element_entity.GetCapture().AddRect("Coverage", *element_entity_coverage,
-                                          {.readonly = true});
-    }
+  if (!clip_stack_->AppendClipCoverage(clip_coverage, element_entity,
+                                       clip_depth_floor,
+                                       global_pass_position)) {
+    return true;
```

Comment:
Maybe store the 1-line comment explaining why this state is terminal

Reply:
Done